modules/javafx.graphics/src/main/java/com/sun/javafx/util/DataURI.java

```diff
@@ -0,0 +1,211 @@
+                if (separator < 0) {
+                    if (i < headers.length - 1) {
+                        throw new IllegalArgumentException("Invalid URI: " + uri);
+                    }
+
+                    base64 = "base64".equalsIgnoreCase(headers[headers.length - 1]);
+                } else {
+                    if (nameValuePairs.isEmpty()) {
+                        nameValuePairs = new HashMap<>();
+                    }
+
+                    nameValuePairs.put(header.substring(0, separator).toLowerCase(), header.substring(separator + 1));
+                }
+            }
+        }
+
+        String data = uri.substring(dataSeparator + 1);
+        Charset charset = Charset.defaultCharset();
+
+        return new DataURI(
+            uri,
+            data,
+            mimeType,
+            mimeSubtype,
+            nameValuePairs,
+            base64,
+            base64 ?
+                Base64.getDecoder().decode(data) :
+                URLDecoder.decode(data.replace("+", "%2B"), charset).getBytes(charset));
+    }
+
+    private final String originalUri;
+    private final String originalData;
+    private final String mimeType, mimeSubtype;
+    private final Map<String, String> parameters;
+    private final boolean base64;
+    private final byte[] data;
+
+    private DataURI(
+            String originalUri,
+            String originalData,
+            String mimeType,
+            String mimeSubtype,
+            Map<String, String> parameters,
+            boolean base64,
+            byte[] decodedData) {
+        this.originalUri = originalUri;
+        this.originalData = originalData;
+        this.mimeType = mimeType;
+        this.mimeSubtype = mimeSubtype;
+        this.parameters = parameters;
+        this.base64 = base64;
+        this.data = decodedData;
+    }
+
+    /**
+     * Returns the MIME type that was specified in the URI.
+     * If no MIME type was specified, returns "text".
+     */
+    public String getMimeType() {
+        return mimeType;
+    }
+
+    /**
+     * Returns the MIME subtype that was specified in the URI.
+     * If no MIME subtype was specified, returns "plain".
+     */
+    public String getMimeSubtype() {
+        return mimeSubtype;
+    }
+
+    /**
+     * Returns the key-value parameter pairs that were specified in the URI.
+     */
+    public Map<String, String> getParameters() {
+        return parameters;
+    }
+
+    /**
+     * Returns whether the data in the URI is Base64-encoded.
+     * If {@code false}, the data is implied to be URL-encoded.
+     */
+    public boolean isBase64() {
+        return base64;
+    }
+
+    /**
+     * Returns the data that is encoded in this URI.
+     * <p>Note that repeated calls to this method will return the same array instance.
+     */
+    public byte[] getData() {
+        return data;
+    }
+
+    @Override
+    public String toString() {
+        if (originalData.length() < 30) {
```

Comment:
I think this should be `< 32`. As it stands, if length is `30` or `31` the data portion of the string will be 31 characters with either 2 or 3 middle chars replaced by `"..."`. In those cases it would be better to just use the original data.